infor-design/enterprise-wc
# 295 - Ids Popup/Modal/Message Improvements

**Explain the details for making this change. What existing problem does the pull request solve?**
This PR makes a handful of changes to the IdsPopup component and some components that depend on the Popup.  These changes include:
- Separates the concept of placement from visibility change in the Popup, for finer-grain control of the lifecycle of Popup-based components.
- Adds `position-style="viewport"` to the Popup specifically for Modal-style centering within the viewport with CSS transforms (replaces `setModalPosition` and JS-based calculation for centering the Modal and fixes one part of #295)
- Changes the way overlays are positioned so they no longer get pushed around by other DOM elements (fixes the other part of #295)
- Replaces `this.state` properties with class-level private properties (per recent dev discussions).
- Make many methods in IdsPopup async compatible (show/hide/place/refreshVisibility).
- Implement new Popup lifecycle into Modal/About/Message/Tooltip components.
- Removes the IdsResizeMixin (per recent dev discussions, it's not needed as ResizeObserver is more performant when used directly)
- Adds the loading of the Resize Observer mock globally to `jest.config.js` to accommodate direct use of ResizeObserver in source code.
- Adds missing/new docs and Typescript defs

**Related github/jira issue (required)**:
Closes #295 

**Steps necessary to review your pull request (required)**:
Pull, build, run, then:

_Test the original issue #295:_
- Open http://localhost:4300/ids-message/visible.html
- When the Message is displayed, the Overlay should be in the correct place and cover the entire viewport.  The Theme Switcher button should not be usable.
- Try refreshing the page several times.  When displayed, the Message should not appear in a different place each time.
- Resize the browser viewport.  The Message should remain centered in the viewport.  The overlay should continue to sit over top of the rest of the page.

_Test other changes:_
All tests should pass.  Also, smoke test the following components to ensure everything is working as expected (specifically Popups opening/closing/being in the correct place):
- http://localhost:4300/ids-popup
- http://localhost:4300/ids-popup/test-sandbox.html
- http://localhost:4300/ids-about
- http://localhost:4300/ids-message
- http://localhost:4300/ids-menu-button
- http://localhost:4300/ids-popup-menu
- http://localhost:4300/ids-tooltip
- http://localhost:4300/ids-tooltip/test-sandbox.html

**Included in this Pull Request**:
